<h1>Project : United Kingdom Road Accident Data Analysis</h1>

<h4>Alley Kyle Regala</h4>

In [1]:
import warnings
warnings.filterwarnings('ignore')

<h1>Import Libraries</h1>

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


<h1>Dataframe</h1>

In [3]:
aksident = pd.read_csv('dataset\\accident_data.csv')

In [4]:
aksident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    05/06/2019  51.506187   
1       200701BS65737           Serious    02/07/2019  51.495029   
2       200701BS66127           Serious    26/08/2019  51.517715   
3       200701BS66128           Serious    16/08/2019  51.495478   
4       200701BS66837            Slight    03/09/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18/02/2022  57.374005   
660675  201091NM01881            Slight    21/02/2022  57.232273   
660676  201091NM01935            Slight    23/02/2022  57.585044   
660677  201091NM01964           Serious    23/02/2022  57.214898   
660678  201091NM02142           Serious    28/02/2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [5]:
aksident_total = aksident['Index'].count()

In [6]:
aksident_total

np.int64(660679)

In [7]:
aksident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [8]:
aksident['Accident_Severity'] = aksident['Accident_Severity'].astype('category')

<h1>Converting Object into DateTime Data Type</h1>

In [9]:
aksident['Accident Date'] = pd.to_datetime(aksident['Accident Date'], dayfirst = True, errors = 'coerce')

In [10]:
aksident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
dtype: object

In [11]:
aksident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

<h1>Filling up missing value</h1>

In [12]:
aksident['Latitude'] = aksident['Latitude'].fillna(aksident['Latitude'].mode()[0])
aksident['Longitude'] = aksident ['Longitude'].fillna(aksident['Longitude'].mode()[0])
aksident['Road_Surface_Conditions'] = aksident['Road_Surface_Conditions'].fillna('unknown Road Surface Conditions')
aksident['Road_Type'] = aksident['Road_Type'].fillna('unknown Road Type')
aksident['Urban_or_Rural_Area'] = aksident['Urban_or_Rural_Area'].fillna('unknown Urban or Rural Area')
aksident['Weather_Conditions'] = aksident['Weather_Conditions'].fillna('unknown Weather Conditions')

In [13]:
aksident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [14]:
aksident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
dtype: object

In [15]:
aksident['Latitude'] = aksident['Latitude'].astype('category')
aksident['Light_Conditions'] = aksident['Light_Conditions'].astype('category')
aksident['District Area'] = aksident['District Area'].astype('category')
aksident['Longitude'] = aksident['Longitude'].astype('category')
aksident['Road_Surface_Conditions'] = aksident['Road_Surface_Conditions'].astype('category')
aksident['Road_Type'] = aksident['Road_Type'].astype('category')
aksident['Urban_or_Rural_Area'] = aksident['Urban_or_Rural_Area'].astype('category')
aksident['Weather_Conditions'] = aksident['Weather_Conditions'].astype('category')
aksident['Vehicle_Type'] = aksident['Vehicle_Type'].astype('category')
aksident['Road_Type'] = aksident['Road_Type'].astype('category')

In [16]:
aksident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                         category
Light_Conditions                 category
District Area                    category
Longitude                        category
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

In [17]:
aksident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1>Extracting date info</h1>

In [18]:
aksident['Year'] = aksident['Accident Date'].dt.year
aksident['Month'] = aksident['Accident Date'].dt.month
aksident['Day'] = aksident['Accident Date'].dt.day
aksident['DayOfWeek'] = aksident['Accident Date'].dt.dayofweek

<h1>UniVariate</h1>

In [19]:
df_2019 = aksident[aksident['Year'] == 2019 ]

In [20]:
df_2019

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
182110       2.01E+12            Slight    2019-12-20  54.985289   
182111       2.01E+12           Serious    2019-12-21  54.984105   
182112       2.01E+12            Slight    2019-12-23  55.166369   
182113       2.01E+12            Slight    2019-12-23  54.995154   
182114       2.01E+12           Serious    2019-12-22  55.104618   

                   Light_Conditions           District Area Longitude  \
0             Darkness - lights lit  Kensington and Chelsea -0.209082   
1                          Daylight  Kensington and Chelsea -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea -0.210215   
3                          Daylight  Kensington and Chelsea -0.202731   
4             Darkness - lights lit  Kensington and Chelsea -0.192487   
...                             ...                     ...       ...   
182110       Darkness - no lighting   Dumfries and Galloway -3.210294   
182111                     Daylight   Dumfries and Galloway -3.193693   
182112       Darkness - no lighting   Dumfries and Galloway -2.992068   
182113        Darkness - lights lit   Dumfries and Galloway -3.058338   
182114                     Daylight   Dumfries and Galloway -3.348857   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
182110                     1                   1            Frost or ice   
182111                     2                   1            Frost or ice   
182112                     1                   1            Frost or ice   
182113                     1                   1             Wet or damp   
182114                     2                   2            Frost or ice   

                 Road_Type Urban_or_Rural_Area          Weather_Conditions  \
0       Single carriageway               Urban          Fine no high winds   
1       Single carriageway               Urban       Raining no high winds   
2        unknown Road Type               Urban  unknown Weather Conditions   
3       Single carriageway               Urban          Fine no high winds   
4        unknown Road Type               Urban  unknown Weather Conditions   
...                    ...                 ...                         ...   
182110  Single carriageway               Rural                       Other   
182111  Single carriageway               Rural                       Other   
182112  Single carriageway               Rural                 Fog or mist   
182113  Single carriageway               Rural          Fine no high winds   
182114  Single carriageway               Rural       Raining no high winds   

                                Vehicle_Type  Year  Month  Day  DayOfWeek  
0                                        Car  2019      6    5          2  
1                                        Car  2019      7    2          1  
2                      Taxi/Private hire car  2019      8   26          0  
3       Bus or coach (17 or more pass seats)  2019      8   16          4  
4                              Other vehicle  2019      9    3          1  
...             

In [21]:
df_2019['Light_Conditions'].value_counts()

Light_Conditions
Daylight                       133042
Darkness - lights lit           35769
Darkness - no lighting          10842
Darkness - lighting unknown      1768
Darkness - lights unlit           694
Name: count, dtype: int64

In [22]:
df_2019['Light_Conditions'].value_counts().sum()

np.int64(182115)

<h1>Using Percentage Difference Calculator</h1>
<h3>Answer : There is a = 31.1419% difference</h3>

<h1>BiVariate</h1>

In [23]:
fine = aksident[aksident['Weather_Conditions'] == 'Fine no high winds']
fine['Vehicle_Type'].value_counts()

Vehicle_Type
Car                                      392668
Van / Goods 3.5 tonnes mgw or under       26877
Bus or coach (17 or more pass seats)      20398
Motorcycle over 500cc                     20245
Goods 7.5 tonnes mgw and over             13589
Motorcycle 125cc and under                12064
Taxi/Private hire car                     10452
Motorcycle over 125cc and up to 500cc      6069
Motorcycle 50cc and under                  6017
Goods over 3.5t. and under 7.5t            4823
Other vehicle                              4450
Minibus (8 - 16 passenger seats)           1552
Agricultural vehicle                       1513
Pedal cycle                                 161
Data missing or out of range                  4
Ridden horse                                  3
Name: count, dtype: int64

<h1>Total of Accident because of "TGIF" Motto</h1>

<h1>MultiVariate</h1>

In [24]:
Severity = aksident.groupby("Accident_Severity").count()

In [25]:
Severity

Index  Accident Date  Latitude  Light_Conditions  \
Accident_Severity                                                      
Fatal                8661           8661      8661              8661   
Serious             88217          88217     88217             88217   
Slight             563801         563801    563801            563801   

                   District Area  Longitude  Number_of_Casualties  \
Accident_Severity                                                   
Fatal                       8661       8661                  8661   
Serious                    88217      88217                 88217   
Slight                    563801     563801                563801   

                   Number_of_Vehicles  Road_Surface_Conditions  Road_Type  \
Accident_Severity                                                           
Fatal                            8661                     8661       8661   
Serious                         88217                    88217      88217   
Slight                         563801                   563801     563801   

                   Urban_or_Rural_Area  Weather_Conditions  Vehicle_Type  \
Accident_Severity                                                          
Fatal                             8661                8661          8661   
Serious                          88217               88217         88217   
Slight                          563801              563801        563801   

                     Year   Month     Day  DayOfWeek  
Accident_Severity                                     
Fatal                8661    8661    8661       8661  
Serious             88217   88217   88217      88217  
Slight             563801  563801  563801     563801

<h1>Insight 1 : Total of accident happening every friday, because of the possible 'tgif' moto of every worker or students.</h1>

In [26]:
friday = aksident[aksident['DayOfWeek'] == 4]

In [27]:
friday.value_counts().sum()

np.int64(97900)

<h4>There is a total number of 97900 accident that happens every friday because of the possible 'tgif motto</h4>

<h1>Insight 2 : Correlation between Number of Casualties & Number of vehicles</h1>
<h3>is it possible that the higher vehicle number involve in a situaion is that there is a higher number of casualties?</h3>

In [29]:
corr = aksident['Number_of_Casualties'].corr(aksident['Number_of_Vehicles'])

In [30]:
corr

np.float64(0.2288888612692756)

<h1>There is no Correlation between Number of Casualties & Number of vehicles</h1>

<h1>Inisight 3 : Place that accident most happens</h1>
<h3>Rural or Urban?</h3>

In [32]:
ruban = aksident['Urban_or_Rural_Area']

In [33]:
ruban.value_counts()


Urban_or_Rural_Area
Urban                          421663
Rural                          238990
unknown Urban or Rural Area        15
Unallocated                        11
Name: count, dtype: int64

<h1>Insight 4 : Total number of accidents in 2022</h1>

In [37]:
total_aksident_2022 = aksident[aksident['Accident Date'].dt.year == 2022]['Accident Date'].count()
total_aksident_2022

np.int64(144419)

In [35]:
ser.value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h1>Insight 5 : Average number of accidents per day</h1>

In [38]:
aksident_per_day = aksident['Accident Date'].groupby(aksident['Accident Date'].dt.day).count().mean()
aksident_per_day

np.float64(21312.225806451614)

<h1>Insight 6 : Types of severity that most happen in all accidents</h1>

In [39]:
Types = aksident['Accident_Severity'].value_counts()
Types

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

<h1>Insight 7 : Most common light condition during accidents</h1>

In [40]:
common_light_condition = aksident['Light_Conditions'].value_counts()
common_light_condition

Light_Conditions
Daylight                       484880
Darkness - lights lit          129335
Darkness - no lighting          37437
Darkness - lighting unknown      6484
Darkness - lights unlit          2543
Name: count, dtype: int64

<h3>The mpst common condition that accident happen is in Daylight</h3>

<h1>Insight 8 : Most weather condition that accident happens</h1>

In [42]:
most_weather = aksident['Weather_Conditions'].value_counts()
most_weather

Weather_Conditions
Fine no high winds            520885
Raining no high winds          79696
Other                          17150
unknown Weather Conditions     14128
Raining + high winds            9615
Fine + high winds               8554
Snowing no high winds           6238
Fog or mist                     3528
Snowing + high winds             885
Name: count, dtype: int64

<h4>'Fine no winds' is the weather condition that accident most happens</h4>

<h1>Insight 9 : Average number of vehicles involved in each accident</h1>

In [43]:
average_vehicles = aksident['Number_of_Vehicles'].mean()
average_vehicles

np.float64(1.8312554205597575)

<h4>There's a 1.83 average number of vehicle that is invilve in each accident</h4>

<h1>Insight 10 :  Most common road surface condition where accidents occurred</h1>

In [44]:
common_surface = aksident['Road_Surface_Conditions'].value_counts()
common_surface

Road_Surface_Conditions
Dry                                447821
Wet or damp                        186708
Frost or ice                        18517
Snow                                 5890
Flood over 3cm. deep                 1017
unknown Road Surface Conditions       726
Name: count, dtype: int64

<h4>Dry road surface is the most road condition that accident mnost happens</h4>

<h1>Insight 11 : Day of the week that has the  highest number of accidents</h1>

In [46]:
day_aksident = aksident['DayOfWeek'].value_counts()
day_aksident

DayOfWeek
5    107178
2     99558
3     99511
4     97900
1     94550
6     89302
0     72680
Name: count, dtype: int64

<h4>Friday (Day 5) recorded the highest number of accidents, with 107,178 incidents.</h4>

<h1>Insight 12 : road type experienced the most accidents happens</h1>

In [47]:
road_type = aksident['Road_Type'].value_counts()
road_type

Road_Type
Single carriageway    492143
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
unknown Road Type       4520
Name: count, dtype: int64

<h4>The most accidents occurred on Single Carriageway roads, with 492,143 recorded incidents. </h4>

<h1>Insight 13 : Do weather condition affect the number of accidents</h1>

In [48]:
weather_aksident = aksident.groupby('Weather_Conditions')['Accident Date'].count().sort_values()
weather_aksident

Weather_Conditions
Snowing + high winds             885
Fog or mist                     3528
Snowing no high winds           6238
Fine + high winds               8554
Raining + high winds            9615
unknown Weather Conditions     14128
Other                          17150
Raining no high winds          79696
Fine no high winds            520885
Name: Accident Date, dtype: int64

<h1>Insight 14 : Average Number of casualties on different road surface conditions</h1>

In [49]:
average_casualties = aksident.groupby('Road_Surface_Conditions')['Number_of_Casualties'].mean()
average_casualties

Road_Surface_Conditions
Dry                                1.334736
Flood over 3cm. deep               1.495575
Frost or ice                       1.348815
Snow                               1.359593
Wet or damp                        1.411064
unknown Road Surface Conditions    1.216253
Name: Number_of_Casualties, dtype: float64

<h1>Insight 15 : Number of Accident that involve Bus vehicle</h1>

In [65]:
horse_aksident = aksident[aksident['Vehicle_Type'] == 'Bus or coach (17 or more pass seats)']
horse_aksident.groupby('Accident_Severity').size()

Accident_Severity
Fatal        325
Serious     3373
Slight     22180
dtype: int64

<h4>'Slight' accident severity is most condition in an accident which involve bus vehicle</h4>

<h1>Insight 16 : Do weather conditions relate to light conditions in road accidents</h1>

In [50]:
weather_condition = aksident.groupby(['Weather_Conditions', 'Light_Conditions'])['Accident Date'].count()
weather_condition

Weather_Conditions          Light_Conditions           
Fine + high winds           Darkness - lighting unknown        59
                            Darkness - lights lit            1852
                            Darkness - lights unlit            51
                            Darkness - no lighting            796
                            Daylight                         5796
Fine no high winds          Darkness - lighting unknown      3683
                            Darkness - lights lit           91951
                            Darkness - lights unlit          1700
                            Darkness - no lighting          24867
                            Daylight                       398684
Fog or mist                 Darkness - lighting unknown        65
                            Darkness - lights lit             921
                            Darkness - lights unlit            37
                            Darkness - no lighting            862
                    

<h4>Most accidents occur in clear weather during daylight, but foggy and rainy conditions increase accidents in low-light conditions. </h4>

<h1>Insight 17 : To know weather conditions, road surface conditions, and accident severity have interaction</h1>

In [51]:
road_severity = aksident.groupby(['Weather_Conditions', 'Road_Surface_Conditions'])['Accident_Severity'].value_counts()
road_severity

Weather_Conditions          Road_Surface_Conditions          Accident_Severity
Fine + high winds           Dry                              Slight               4162
                                                             Serious               758
                                                             Fatal                 103
                            Flood over 3cm. deep             Slight                 14
                                                             Fatal                   1
                                                                                  ... 
unknown Weather Conditions  Wet or damp                      Serious               169
                                                             Fatal                  29
                            unknown Road Surface Conditions  Slight                550
                                                             Serious                61
                                                   

<h1>Insight 18 : To know road type, light conditions, and accident severity relate to each other</h1>

In [52]:
light_severity = aksident.groupby(['Road_Type', 'Light_Conditions'])['Accident_Severity'].value_counts()
light_severity

Road_Type          Light_Conditions             Accident_Severity
Dual carriageway   Darkness - lighting unknown  Slight                 805
                                                Serious                 93
                                                Fatal                    7
                   Darkness - lights lit        Slight               17025
                                                Serious               2762
                                                                     ...  
unknown Road Type  Darkness - no lighting       Serious                 33
                                                Fatal                    1
                   Daylight                     Slight                3098
                                                Serious                349
                                                Fatal                   22
Name: count, Length: 90, dtype: int64

<h1>Insight 19 : Corelation between vehicle type, urban/rural area, and accident severity</h1>

In [53]:
vehicle_area_severity = aksident.groupby(['Vehicle_Type', 'Urban_or_Rural_Area'])['Accident_Severity'].value_counts()
vehicle_area_severity

Vehicle_Type                         Urban_or_Rural_Area          Accident_Severity
Agricultural vehicle                 Rural                        Slight                540
                                                                  Serious               121
                                                                  Fatal                  14
                                     Unallocated                  Fatal                   0
                                                                  Serious                 0
                                                                                       ... 
Van / Goods 3.5 tonnes mgw or under  Urban                        Serious              2707
                                                                  Fatal                 161
                                     unknown Urban or Rural Area  Fatal                   0
                                                                  Serious               

<h1>Insight 20 : Number of accident that involve Car vehicles</h1>

In [60]:
horse_aksident = aksident[aksident['Vehicle_Type'] == 'Car']
horse_aksident.groupby('Accident_Severity').size()

Accident_Severity
Fatal        6577
Serious     66461
Slight     424954
dtype: int64

<h4>'Slight' accident severity is most condition in an accident which involve car vehicle</h4>
